gyp: fix node build failure with long build directory path on Linux

NeoRaider opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j5 test` passes (I saw some test failures when running the build/tests in a very long path, mostly EADDRINUSE for socket files; no test failures with this patch in shorter paths)
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
- build
##### Description of change

<!-- Provide a description of the change below this comment. -->

Some build targets (especially the embedded OpenSSL) would use very long
ar command lines (in the case of OpenSSL, containing more than 600 object
files). This would often exceed the argument list limit for common stack
size settings when building in directories with long paths.

Fix this by writing the object list to a file first and referring to it in
the ar command.

All branches are affected, I found reports of this issue going back to node v0.8.x.

Based-on-patch-by: Amy Fong amy.fong@windriver.com
Fixes: https://github.com/nodejs/node/issues/9137

NeoRaider commented:
The longer I look at the patch, the more issues I find in it...

There's code in the makefile generator to handle spaces in paths, is this supported for building node? The simple "xargs" definition I (mostly) took from the patch I found doesn't handle this; if correct handling of spaces is desired, I try to fix this tomorrow.

bnoordhuis commented:
Don't change deps/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py, that would get blown away on the next npm upgrade.

Did you have any luck with upping `ulimit -s`?

NeoRaider commented:
I've removed the deps/npm/node_modules change and added a note about missing space handling.

I couldn't get it to work upping `ulimit -s`, I went as far as setting it to 1048576.

gibfahn commented:
Spaces in the path are definitely breaking the build for me:

```
/tmp/gib/space node $ make -j8
make -C out BUILDTYPE=Release V=1
make[1]: Entering directory `/tmp/gib/space node/out'
Makefile:263: *** mixed implicit and normal rules.  Stop.
make[1]: Leaving directory `/tmp/gib/space node/out'
make: *** [node] Error 2
```

feckert commented:
@NeoRaider I get faced now with the same problem on my build env! 
What can i do to get this working (despite increasing the ulimit) with master/lede-17.01?
Are there any plans to get this patch into "lede" "master/17.01" (As long as "lede" will update node to a new version with this patch)?

builds for x86_64 has no such issue the different  pathes are:
> /home/bbslave/buildSlave/lede_dev_lantiq_xrx200
> /home/bbslave/buildSlave/lede_dev_x86_64

NeoRaider commented:
@feckert The nodejs issue tracker is certainly not the right place to ask for inclusion of a patch in OpenWrt/LEDE.

Nevertheless... any chance we can get this PR accepted into node?

jasnell commented:
Ping @nodejs/build ... does this make sense to do?

bnoordhuis commented:
Also @nodejs/node-gyp - if we accept this patch, it makes sense to land it there as well.

I like the idea of this pull request but I don't love the execution.  The risk/reward ratio is not quite there: 60 lines of hard to understand code in exchange for fixing (or rather: working around) an obscure issue.  I'm also not completely convinced it's necessary, see https://github.com/nodejs/node/issues/9137#issuecomment-254299940.

feckert commented:
I am not a contributer to node or so, but I am working on the luci-ng webgui for lede which uses node as a build env. And without this patch i am unable to compile my working env and . I have added this patch with minor modification to apply against node version on lede 17.01 and every think is fine for now. I can compile my sources. I hope the issue get fixed with this patch or an other solution.

fhinkel commented:
There hasn't been any activity here. I'm closing this. Feel free to reopen (or ping a collaborator) if I closed this in error.
